closure_captures.md

```diff
@@ -0,0 +1,319 @@
+
+#### [`TypeckTables::closure_captures`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.TypeckTables.html#structfield.closure_captures)
+
+Provides list of root variables that are completely or partially captured by the closure, and maps them to an `UpvarId`.
+`UpvarId` contains the closure definition id and the hir id of the root variable.
+
+```
+closure_captures[c1] =  { hir_id_1 : UpvarId(def_id_c1, hir_id_1) }
+closure_captures[c2] =  { hir_id_3 : UpvarId(def_id_c2, hir_id_3) }
+closure_captures[c3] =  {}
+closure_captures[c_nest] =  { hir_id_2: UpvarId(def_id_c_nest, hir_id_2) }
+closure_captures[c4] =  { hir_id_2: UpvarId(def_id_c_4, hir_id_2), hir_id_1: UpvarId(def_id_c_4, hir_id_1) }
+closure_captures[c5] =  { hir_id_11 : UpvarId(def_id_c5, hir_id_11) }
+closure_captures[c6] =  { hir_id_11 : UpvarId(def_id_c6, hir_id_11) }
+```
+
+#### [`TypeckTables::upvar_capture_map`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.TypeckTables.html#structfield.upvar_capture_map)
+
+For given scope, maps all the `UpvarId` to capture kind, i.e. by value, or mutabale/immutable reference.
+
+Assume that `c1 ... c6` are within the same scope, then for that scope:
+```
+upvar_capture_map = {
+    UpvarId(def_id_c1, hir_id_1) : ByRef(ImmBorrow),
+    UpvarId(def_id_c2, hir_id_3) : ByRef(ImmBorrow),
+    UpvarId(def_id_c_nest, hir_id_2) : ByRef(ImmBorrow),
+    UpvarId(def_id_c_4, hir_id_2) : ByRef(MutBorrow),
+    UpvarId(def_id_c_4, hir_id_1) : ByValue,
+    UpvarId(def_id_c5, hir_id_11) : ByRef(MutBorrow),
+    UpvarId(def_id_c6, hir_id_11) : ByValue,
+}
+```
+
+#### [`Place`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_typeck/mem_categorization/struct.Place.html)
+
+Represents how memory is accessed.
+
+Eg: `p3.x`, we can express that the base root variable is `p3` and then we get the value by accessing Field(x).
+
+If the [`base`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_typeck/mem_categorization/struct.Place.html#structfield.base)
+field is `PlaceBase::Upvar` we know that the access is based on a captured variable.
+
+### Generating this information
+
+To generate `upvars_mentioned`, `closure_captures`, `upvar_capture_map` we need to iterate the CFG, to see how variables are being captures and used.
+
+[`rustc_hir::instravisit`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/intravisit/index.html)
+provides a visitor/walker that can traverse the tree in execution order.
+We can use it and override `visit_*` methods, to have a walker that meets our needs.
+
+
+#### [`upvars_mentioned`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/arena/struct.Arena.html#structfield.upvars_mentioned)
+Generated in [`librustc_passes`](https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc_passes/upvars.rs.html#82).
+
+Here we implement the instravisit visitor that can generate a list of local root variables by looking at `PatKind::Binding` (**TODO: explain**).
```

Comment:
I'm not sure how PatKind::Binding works and how using a variable ourside of pattern matching relates to a Pattern

Reply:
I don't really understand what you're referring to here, we don't seem to be missing bindings in patterns in the `upvars_mentioned` code in question -- i.e., [in this code](https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc_passes/upvars.rs.html#82). We seem to be visiting paths -- am I missing something?

Reply:
https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc_passes/upvars.rs.html#53

Reply:
I see. This code needs more comments. Anyway, that is collecting all the local variables declared within the function. A "binding" could be the `x` in `let x = ..` or the `x` in a more complex case like `let (x, _) = ...` or the `x` from `match option { Some(x) => .. }`. In all cases, it's a local variable being introduced.

So basically it's computing:

* What is the set of local variable ids created within the closure (call it C)
* What is the set of local variable ids referenced by some code in the closure (call it U)

If you take the set difference `U \ C` that must be the "upvars" -- i.e., the variables the closure *uses* but did not *define* (because they were defined in the parent).

Reply:
So I was thinking pattern was what is used inside a match block, but it seems to be much more than that.

Leaving this here, in case any one else is confused
https://doc.rust-lang.org/book/ch18-03-pattern-syntax.html